### ISSUE TITLE
formal description of the keywords in the demes yaml format

### ISSUE THREAD

**Issue by grahamgower** [grahamgower/demes#57]
Along with #56, we want a precise and complete description of all the keywords (`description`, `doi`, `demes`, `epoch`, etc), their scope, and semantics.

**Comment by apragsdale** [comment]
From today's call, it's clear we also need a precise definition for what `time` means, especially for discrete forward simulators.

E.g., if a split occurs at time `T`, does that mean the child demes come into existence at time `T+1`, and the parent deme's last generation "alive" is at time `T`? Or should that be shifted forward one generation, so that the child demes show up at time `T`.

I think we settled on the latter. As an illustration, take some event, say a split, that we specify to occur at time 0 (looking backward in time, meaning in the generation that sampling takes place). We want the child demes from the split to be present at the time of sampling at the end of the simulation, and for the parent to exist up to just one generation before the sampling time and then no longer exists at the sampling time. If the event were shifted one generation forward, the child demes would never show up.

**Comment by apragsdale** [comment]
I just realized that this illustration isn't possible though, because the child demes would have `time_span`s of 0, which isn't permitted. So this exact example isn't great, but I think the idea that events occur so that the split/merged/admixed demes are present _starting at_ the specified `time` is the way to go still.

Possible I'm still confused on this point though... correct me if I'm off here.

**Comment by grahamgower** [comment]
Doesn't it work better if we define a deme's lifespan to be the half-open interval `(start_time, end_time]`? Demes can have `start_time=inf` and/or `end_time=0`.

```yaml
description: deme A terminates and becomes deme B
time_units: generations
demes:
  A:
    initial_size: 1000
    end_time: 100
  B:
    initial_size: 1000
    ancestors: A
```

```python
>>> print(*demes.load("/tmp/a.yml").demes, sep="\n")
Deme(id='A', description=None, ancestors=None, proportions=None, epochs=[Epoch(start_time=inf, end_time=100, initial_size=1000, final_size=1000, size_function='constant', selfing_rate=None, cloning_rate=None)], selfing_rate=None, cloning_rate=None)
Deme(id='B', description=None, ancestors=['A'], proportions=[1.0], epochs=[Epoch(start_time=100, end_time=0, initial_size=1000, final_size=1000, size_function='constant', selfing_rate=None, cloning_rate=None)], selfing_rate=None, cloning_rate=None)
```

**Comment by jeromekelleher** [comment]
I'm all for half-open intervals.

**Comment by apragsdale** [comment]
Yeah, you're both right - half-open intervals make a lot of sense and it settles the question about what `when` means for a split or a size change or other event. I have definitely been overthinking it and just confusing myself :)
